### ISSUE TITLE
Multiple users subscribe with different fields. 

### ISSUE THREAD

**Issue by hansololai** [apollographql/apollo-server#1743]
- [ x] feature

Suppose I have a subscription definition
```
type Subscription{
  userUpdated(id:int): User
}
```
and user have id, name, birthday, 3 fields
The client can make a subscription like this:
```
Subscription{
  userUpdated(id:1){
    id
    name
  }
}
```

Then In the places where I have resolvers for mutation,
```
// Inside mutation resolver, at the right place
        pubsub.publish('userUpdated', {
          userUpdated: {
            id,
            name,
            birthday
          },
        });
```

My question is, what if the client was subscribed to a subset of the field, say only the "id" field.  The server have no idea what the client was subscribed to, but will return the entire user object. Also, what if the user subscribes to a nested object too? If user has a contact object.  For example
```
Subscription{
  userUpdated(id:1){
     name
    contact:{
       phoneNumber
    }
  }
}
```

So if the apollo server publish {id, name, birthday}, then apollo server will receive the data, but because it didn't have contact, it throws warning, and worse, it is "smartly" writing the cache of User with id=1 and remove the contact field. 

In order to keep track of this, I will need to have a storage object that stores each subscription client, and stores their query. So if two clients, one client subscribe to id, name field, another client subscribe to  birthday  and contact field.  Then I will need to store these queries in some storage, 
```
subscriptionQueries={userUpdated:[//all queries here]}
```
And in the mutation resolver:
```
// Inside mutation resolver, at the right place
const data ={}    
subscriptionQueries.userUpdated.forEach(q=>{
      // Merge all the fields required by each client.  In this case
      // would be id, name, birthday and contact
      data.id = ...
      ...
    });
        pubsub.publish('userUpdated', {
          userUpdated: data,
        });
```

This is a lot of work and I I feel like if a lot of subscriptions are made, the mutation resolver could get slow. I don't have any idea how many subscription is too many, and I am also curious if anyone knows if merging queries, keeping track of all subscription queries, will that take too much memory, too much cpu? etc. 

Thanks

**Comment by jsegaran** [comment]
I'll close this since this doesn't appear to be a bug with Apollo Server, but rather a question about it.

Rather than asking it here in GitHub Issues — where efforts are focused on fixing bugs and adding new features — I'd ask that you take this question to the _Apollo Server_ channel within [the Apollo community on Spectrum.chat](https://spectrum.chat/apollo) where there are community members who might be able to relate to a similar problem, or might be able to help you out more interactively.  Thanks for your understanding!

**Comment by hansololai** [comment]
I asked this question a long time ago, after that I have been reading the source code, asking in spectrum, (not very helpful in this particular case), After reading some source code, and I have discovered  a way to resolve this issue. I made a package in my repo, the code is pretty simply to read through. 
https://github.com/hansololai/apollo_subscription_boilerplate

consider this issue is resolved.
